Source/Dafny/DafnyAst.cs

```diff
@@ -325,30 +325,67 @@ public class BuiltIns
       return new ArrowType(f.tok, atd, f.Formals.ConvertAll(arg => Resolver.SubstType(arg.Type, typeMap)), Resolver.SubstType(f.ResultType, typeMap));
     }
 
-    public TupleTypeDecl TupleType(IToken tok, int dims, bool allowCreationOfNewType) {
+    private object MakeTupleKey(List<bool> isGhost, int dims) {
+      if (dims == 0) {
+        return 0;
+      } else {
+        var g = isGhost[dims - 1];
+        return Tuple.Create(g, MakeTupleKey(isGhost, dims - 1));
+      }
+    }
+
+    public TupleTypeDecl TupleType(IToken tok, int dims, bool allowCreationOfNewType, List<bool> ghostArguments = null) {
       Contract.Requires(tok != null);
       Contract.Requires(0 <= dims);
       Contract.Ensures(Contract.Result<TupleTypeDecl>() != null);
 
       TupleTypeDecl tt;
+      if (ghostArguments == null) {
+        ghostArguments = new bool[dims].ToList().ConvertAll(_ => false);
```

Comment:
Consider using `new bool[dims].Select(_ => false).ToList();` instead.

My impression is that `ConvertAll` has been replaced by `Select` in practice, so most C# devs won't know `ConvertAll` and it's better to avoid it. [More information](https://stackoverflow.com/a/1571843/93197)

Reply:
`ConvertAll` has the advantage that it returns a `List`, so there's no need to follow it with `.ToList()` (and the implementation then immediately knows how much storage to allocate for the `List` to be created).